lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp

```diff
@@ -382,227 +383,289 @@ const CompilerType &SwiftLanguageRuntimeImpl::GetBoxMetadataType() {
 
-    auto format_data = [](auto dest, auto size) {
+  LLDB_LOGV(log, "[MemoryReader] asked to read string data at address {0x}",
+            address.getAddressData());
+
+  Address addr = Address(address.getAddressData());
+  if (auto resolved_addr = tryToConvertFromMappedAddressIntoRealAddress(
+          address.getAddressData()))
+    addr = *resolved_addr;
+
+  Target &target(m_process.GetTarget());
+  Status error;
+  target.ReadCStringFromMemory(addr, dest, error);
+  if (error.Success()) {
+    auto format_string = [](const std::string &dest) {
       StreamString stream;
-      for (uint64_t i = 0; i < size; i++) {
-        stream.PutHex8(dest[i]);
-        stream.PutChar(' ');
+      for (auto c : dest) {
+        if (c >= 32 && c <= 127) {
+          stream << c;
+        } else {
+          stream << "\\0";
+          stream.PutHex8(c);
+        }
       }
       return stream.GetData();
     };
-    LLDB_LOGV(log, "[MemoryReader] memory read returned data: {0}",
-              format_data(dest, size));
-
+    LLDB_LOGV(log, "[MemoryReader] memory read returned data: \"{0}\"",
+              format_string(dest));
     return true;
+  } else {
+    LLDB_LOGV(log, "[MemoryReader] memory read returned error: {0}",
+              error.AsCString());
+    return false;
   }
+}
 
-  bool readString(swift::remote::RemoteAddress address,
-                  std::string &dest) override {
-    Log *log(GetLogIfAllCategoriesSet(LIBLLDB_LOG_TYPES));
+void LLDBMemoryReader::pushLocalBuffer(uint64_t local_buffer,
+                                       uint64_t local_buffer_size) {
+  lldbassert(!m_local_buffer);
+  m_local_buffer = local_buffer;
+  m_local_buffer_size = local_buffer_size;
+}
 
-    LLDB_LOGV(log, "[MemoryReader] asked to read string data at address {0:x}",
-              address.getAddressData());
+void LLDBMemoryReader::popLocalBuffer() {
+  lldbassert(m_local_buffer);
+  m_local_buffer.reset();
+  m_local_buffer_size = 0;
+}
 
-    Target &target(m_process.GetTarget());
-    Address addr(address.getAddressData());
-    Status error;
-    target.ReadCStringFromMemory(addr, dest, error);
-    if (error.Success()) {
-      auto format_string = [](const std::string &dest) {
-        StreamString stream;
-        for (auto c : dest) {
-          if (c >= 32 && c <= 127) {
-            stream << c;
-          } else {
-            stream << "\\0";
-            stream.PutHex8(c);
-          }
-        }
-        return stream.GetData();
-      };
-      LLDB_LOGV(log, "[MemoryReader] memory read returned data: \"{0}\"",
-                format_string(dest));
-      return true;
-    } else {
-      LLDB_LOGV(log, "[MemoryReader] memory read returned error: {0}",
-                error.AsCString());
-      return false;
-    }
+uint64_t LLDBMemoryReader::addModuleToAddressMap(ModuleSP module) {
+  // We set the highest bit to tell that the address is one mapped by us when
+  // doing memory reads later on.
```

Comment:
Should we cross-check this bit (in an assertions) against the various SWIFT_ABI_X86_64_SWIFT_SPARE_BITS_MASK constants to ensure this is safe?